python_modules/dagster/dagster/_core/definitions/job_definition.py

```diff
@@ -918,7 +918,13 @@ def _copy(self, **kwargs: Any) -> "JobDefinition":
             resource_defs=dict(self.resource_defs),
             executor_def=self._executor_def,
             logger_defs=self._loggers,
-            config=self._config_mapping or self._partitioned_config or self._run_config,
+            # One of the following is true:
+            #   _config_mapping is set, and _partitioned_config and _run_config are None
+            #   _partitioned_config is set, and _config_mapping and _run_config are None
+            #   _run_config is set and _config_mapping is built from it using _config_mapping_with_default_value (and _partitioned_config is None)
+            # In the last case, we want to copy the _run_config, not the _config_mapping, since a new _config_mapping will be built
+            # from the _run_config in the constructor
+            config=self._run_config or self._config_mapping or self._partitioned_config,
```

Comment:
So this seems very hard to get right.

Why don't we instead just add another variable `self._config_to_copy` that we set alongside the rest of this on line 218. Then there is one place where we have to set the logic, and it will be self-evident from the codepath what the cases are rather than encoding it into a comment like this which is likely go out of date at some point.

Reply:
seems reasonable - updated